File: lib/performanceCache.py
"""
Performance Cache Library

Provides caching functionality for the performance-v2 API endpoint.
Cache key includes context + date range: weekly, monthly, yearly, custom-START-END
"""

import json
import os
import threading
from datetime import datetime, date
from typing import Dict, Any, Optional

# Cache storage
_performance_cache: Dict[str, Dict[str, Any]] = {}
_cache_lock = threading.Lock()

# Cache file path
CACHE_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'data')
CACHE_FILE = os.path.join(CACHE_DIR, 'performance_cache.json')


def generate_cache_key(context: str, start_date: Optional[str] = None, end_date: Optional[str] = None, month_offset: int = 0) -> str:
    """
    Generate cache key based on context and date range.
    
    Key formats:
    - weekly → weekly context
    - monthly-YYYY-MM → monthly context with offset
    - yearly → yearly context
    - custom-YYYY-MM-DD-YYYY-MM-DD → custom date range
    """
    if context == 'weekly':
        return 'weekly'
    elif context == 'monthly':
        # Calculate the month based on offset
        today = date.today()
        target_month = today.month - month_offset
        target_year = today.year
        
        while target_month <= 0:
            target_month += 12
            target_year -= 1
        
        return f'monthly-{target_year}-{target_month:02d}'
    elif context == 'yearly':
        return 'yearly'
    elif context == 'custom' and start_date and end_date:
        return f'custom-{start_date}-{end_date}'
    else:
        return context


def get_performance_cache(key: str) -> Optional[Dict[str, Any]]:
    """
    Get cached performance data for the given key.
    
    Args:
        key: Cache key (weekly, monthly-YYYY-MM, yearly, custom-YYYY-MM-DD-YYYY-MM-DD)
    
    Returns:
        Cached data dict or None if not found
    """
    with _cache_lock:
        return _performance_cache.get(key)


def set_performance_cache(key: str, data: Dict[str, Any]) -> None:
    """
    Store performance data in cache.
    
    Args:
        key: Cache key
        data: Performance data to cache
    """
    with _cache_lock:
        _performance_cache[key] = {
            'data': data,
            'timestamp': datetime.now().isoformat()
        }


def clear_performance_cache() -> None:
    """
    Clear all performance cache entries.
    Called when trade events are added/deleted.
    """
    with _cache_lock:
        _performance_cache.clear()


def load_performance_cache_from_file() -> None:
    """Load cache from persistent file if available."""
    if os.path.exists(CACHE_FILE):
        try:
            with open(CACHE_FILE, 'r') as f:
                _performance_cache.update(json.load(f))
        except (json.JSONDecodeError, IOError):
            pass  # Start with empty cache if file is invalid


def save_performance_cache_to_file() -> None:
    """Save cache to persistent file."""
    try:
        with open(CACHE_FILE, 'w') as f:
            json.dump(_performance_cache, f, indent=2)
    except IOError:
        pass  # Silently fail if unable to write


def get_quote_timestamp() -> str:
    """
    Get the quote timestamp from quote cache.
    
    Returns:
        ISO format timestamp string from quote cache
    """
    quote_cache_file = os.path.join(CACHE_DIR, 'quote_cache.json')
    try:
        if os.path.exists(quote_cache_file):
            with open(quote_cache_file, 'r') as f:
                quote_data = json.load(f)
                return quote_data.get('lastQuoteUpdate', datetime.now().isoformat())
    except (json.JSONDecodeError, IOError):
        pass
    return datetime.now().isoformat()

File: lib/performanceCache.ts
/**
 * Performance Cache Library
 * 
 * Provides caching functionality for the performance-v2 API endpoint.
 * Cache key includes context + date range: weekly, monthly, yearly, custom-START-END
 */

import fs from 'fs';
import path from 'path';

interface CacheEntry {
  data: PerformanceData;
  timestamp: string;
}

interface PerformanceData {
  context: string;
  dateRange: {
    start: string | null;
    end: string | null;
  };
  summary: {
    totalRealizedPl: number;
    winRate: number;
    totalTrades: number;
  };
  tickerBreakdown: TickerBreakdown[];
  openPositions?: OpenPosition[];
  equityCurve?: EquityPoint[];
}

interface TickerBreakdown {
  ticker: string;
  realizedPl: number;
  wins: number;
  losses: number;
  winRate: number;
}

interface OpenPosition {
  ticker: string;
  // Add other position fields as needed
}

interface EquityPoint {
  date: string;
  value: number;
}

type CacheStore = Record<string, CacheEntry>;

const CACHE_DIR = path.join(process.cwd(), 'data');
const CACHE_FILE = path.join(CACHE_DIR, 'performance_cache.json');

// In-memory cache
let cache: CacheStore = {};

/**
 * Get Monday of the week containing the given date
 */
function getMonday(date: Date): Date {
  const d = new Date(date);
  const day = d.getDay();
  const diff = d.getDate() - day + (day === 0 ? -6 : 1); // Adjust when day is Sunday
  return new Date(d.setDate(diff));
}

/**
 * Generate cache key based on context and date range
 * 
 * Key formats:
 * - weekly → weekly context
 * - monthly-YYYY-MM → monthly context with offset
 * - yearly → yearly context
 * - custom-YYYY-MM-DD-YYYY-MM-DD → custom date range
 */
export function generateCacheKey(
  context: string, 
  startDate?: string, 
  endDate?: string, 
  monthOffset: number = 0,
  weekOffset: number = 0
): string {
  if (context === 'weekly') {
    if (weekOffset !== 0) {
      // Calculate the target week start date based on weekOffset
      const now = new Date();
      const currentWeekStart = getMonday(now);
      const targetWeekStart = new Date(currentWeekStart.getTime() - weekOffset * 7 * 24 * 60 * 60 * 1000);
      return `weekly-${targetWeekStart.toISOString().split('T')[0]}`;
    }
    return 'weekly';
  } else if (context === 'monthly') {
    const today = new Date();
    const targetMonth = today.getMonth() - monthOffset;
    const targetYear = today.getFullYear() + Math.floor(targetMonth / 12);
    const adjustedMonth = ((targetMonth % 12) + 12) % 12 + 1;
    return `monthly-${targetYear}-${adjustedMonth.toString().padStart(2, '0')}`;
  } else if (context === 'yearly') {
    return 'yearly';
  } else if (context === 'custom' && startDate && endDate) {
    return `custom-${startDate}-${endDate}`;
  }
  return context;
}

/**
 * Get cached performance data for the given key
 */
export function getPerformanceCache(key: string): PerformanceData | null {
  const entry = cache[key];
  if (entry) {
    return entry.data;
  }
  return null;
}

/**
 * Store performance data in cache
 */
export function setPerformanceCache(key: string, data: PerformanceData): void {
  cache[key] = {
    data,
    timestamp: new Date().toISOString()
  };
  saveCacheToFile();
}

/**
 * Clear all performance cache entries
 * Called when trade events are added/deleted
 */
export function clearPerformanceCache(): void {
  cache = {};
  saveCacheToFile();
}

/**
 * Load cache from persistent file if available
 */
export function loadCacheFromFile(): void {
  try {
    if (fs.existsSync(CACHE_FILE)) {
      const fileContent = fs.readFileSync(CACHE_FILE, 'utf-8');
      cache = JSON.parse(fileContent);
    }
  } catch (error) {
    console.error('Failed to load performance cache:', error);
    cache = {};
  }
}

/**
 * Save cache to persistent file
 */
function saveCacheToFile(): void {
  try {
    if (!fs.existsSync(CACHE_DIR)) {
      fs.mkdirSync(CACHE_DIR, { recursive: true });
    }
    fs.writeFileSync(CACHE_FILE, JSON.stringify(cache, null, 2));
  } catch (error) {
    console.error('Failed to save performance cache:', error);
  }
}

/**
 * Get the quote timestamp from quote cache
 */
export function getQuoteTimestamp(): string {
  const quoteCacheFile = path.join(CACHE_DIR, 'quote_cache.json');
  try {
    if (fs.existsSync(quoteCacheFile)) {
      const fileContent = fs.readFileSync(quoteCacheFile, 'utf-8');
      const quoteData = JSON.parse(fileContent);
      return quoteData.lastQuoteUpdate || new Date().toISOString();
    }
  } catch (error) {
    console.error('Failed to read quote cache:', error);
  }
  return new Date().toISOString();
}

// Initialize cache on module load
loadCacheFromFile();

export type { PerformanceData, CacheEntry, TickerBreakdown, OpenPosition, EquityPoint };
